smart-contracts/contracts/mixins/MixinPurchase.sol

```diff
@@ -80,7 +84,10 @@ contract MixinPurchase is
     );
 
     // We explicitly allow for greater amounts of ETH to allow 'donations'
+    // Security: after state changes to minimize risk of re-entrancy
+    uint pricePaid = _chargeAtLeast(netPrice);
+
     // Security: last line to minimize risk of re-entrancy
-    _chargeAtLeast(netPrice);
+    _onKeySold(_recipient, _referrer, pricePaid, _data);
```

Comment:
Isn't there a re-entrancy risk here if the hook calls back `purchase`?

Reply:
It is possible to reenter via the hook.  However that would be fine.

Reentrancy becomes problematic when some state changes for the transaction are not complete before the external call is made.  Here we were able to make it the last line of the purchase function which makes this clearly safe.  Any reentrancy would be on-par with simply calling `purchase` again in a future tx.

Reply:
Yup, agreed that this looks fine as it follows the "checks, effects, interactions" pattern.